### PR TITLE
Projections analysis from query plan

### LINKED ISSUE #44963: DB::Exception: Cannot find column `count(id_with)` in source stream
How to reproduce:
```sql
:) create table test(day Date, id UInt32) engine=MergeTree partition by day order by tuple()                                                    
                                                                                                                                                                                                                    
CREATE TABLE test                                                                                                                                                                                                   
(                                                                                                                                                                                                                   
    `day` Date,                                                                                                                                                                                                     
    `id` UInt32                                                                                                                                                                                                     
)                                                                                                                                                                                                                   
ENGINE = MergeTree                                                                                                                                                                                                  
PARTITION BY day                                                                                                                                                                                                    
ORDER BY tuple()                                                                                                                                                                                                    
                                                                                                                                                                                                                    
Query id: a81e4c06-cbc1-4e18-b1dd-1d68cc30a630                                                                                                                                                                      
                                                                                                                                                                                                                    
Ok.                                                                                                                                                                                                                 
                                                                                                                                                                                                                    
0 rows in set. Elapsed: 0.011 sec.           

:) insert into test select toDate('2023-01-05') AS day, number from numbers(10);

INSERT INTO test SELECT
    toDate('2023-01-05') AS day,
    number
FROM numbers(10)

Query id: 1ee62a35-6c91-4797-9a6e-08518f34995b

Ok.

0 rows in set. Elapsed: 0.002 sec.

 :) with toUInt64(id) as id_with select day, count(id_with)  from test where day >= '2023-01-01' group by day limit 1000

WITH toUInt64(id) AS id_with
SELECT
    day,
    count(id_with)
FROM test
WHERE day >= '2023-01-01'
GROUP BY day
LIMIT 1000

Query id: 362bf093-d693-4655-9d2c-a3e4b687708c


0 rows in set. Elapsed: 0.013 sec. 

Received exception from server (version 22.13.1):
Code: 8. DB::Exception: Received from localhost:9000. DB::Exception: Cannot find column `count(id_with)` in source stream, there are only columns: [day]. (THERE_IS_NO_COLUMN)
```

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Re-implement projection analysis on top of query plan. Added setting `query_plan_optimize_projection=1` to switch between old and new version. Fixes #44963

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 1]
```diff
@@ -1076,24 +1058,7 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
...
-                    size_t total_rows_estimate = current_rows_estimate + prev_total_rows_estimate;
-                    if (query_info.limit > 0 && total_rows_estimate > query_info.limit)
-                    {
-                        total_rows_estimate = query_info.limit;
-                    }
-                    limits.check(total_rows_estimate, 0, "rows (controlled by 'max_rows_to_read' setting)", ErrorCodes::TOO_MANY_ROWS);
-                    leaf_limits.check(
-                        total_rows_estimate, 0, "rows (controlled by 'max_rows_to_read_leaf' setting)", ErrorCodes::TOO_MANY_ROWS);
-                }
-
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1094]: This logic is moved to `ReadFromMergeTree` step.
We can't skip PK analysis early now, but I suppose it's ok.

File: tests/queries/0_stateless/01710_normal_projection_fix1.sql [hunk 2]
```diff
@@ -7,11 +7,14 @@ insert into t values (1, 2);
 alter table t add projection x (select * order by j);
 
 insert into t values (1, 4);
+insert into t values (1, 5);
```
Comment (KochetovNicolai) [inline tests/queries/0_stateless/01710_normal_projection_fix1.sql:10]: This is needed to make reading from projection preferable.

File: tests/queries/0_stateless/01710_projection_aggregation_in_order.sql [hunk 3]
```diff
@@ -21,7 +21,7 @@ ENGINE = MergeTree
 ORDER BY (key, ts);
 
 INSERT INTO normal SELECT
-    1,
+    number,
```
Comment (KochetovNicolai) [inline tests/queries/0_stateless/01710_projection_aggregation_in_order.sql:24]: This is needed to make reading from projection preferable.

File: tests/queries/0_stateless/01710_projections_in_distributed_query.sql [hunk 4]
```diff
@@ -1,5 +1,7 @@
 -- Tags: distributed
 
+set enable_memory_bound_merging_of_aggregation_results=0;
```
Comment (KochetovNicolai) [inline tests/queries/0_stateless/01710_projections_in_distributed_query.sql:3]: Projections does not work with memory bound merging. We can fix it later.

File: tests/queries/0_stateless/02319_no_columns_in_row_level_filter.reference [hunk 5]
```diff
@@ -1,4 +1,4 @@
-1000000
+0
```
Comment (KochetovNicolai) [inline tests/queries/0_stateless/02319_no_columns_in_row_level_filter.reference:1]: This was just wrong. Fixed.

File: tests/queries/0_stateless/02521_cannot-find-column-in-projection.sql [hunk 6]
```diff
@@ -1,3 +1,3 @@
 create table test(day Date, id UInt32) engine=MergeTree partition by day order by tuple();
 insert into test select toDate('2023-01-05') AS day, number from numbers(10);
-with toUInt64(id) as id_with select day, count(id_with)  from test where day >= '2023-01-01' group by day limit 1000; -- { serverError NOT_FOUND_COLUMN_IN_BLOCK }
+with toUInt64(id) as id_with select day, count(id_with)  from test where day >= '2023-01-01' group by day limit 1000;
```
Comment (KochetovNicolai) [inline tests/queries/0_stateless/02521_cannot-find-column-in-projection.sql:3]: Now it just works.

File: tests/queries/1_stateful/00172_early_constant_folding.sql [hunk 7]
```diff
@@ -1,3 +1,4 @@
 -- Tags: no-parallel-replicas
 
+set max_threads=10;
```
Comment (KochetovNicolai) [inline tests/queries/1_stateful/00172_early_constant_folding.sql:3]: Set the number of threads to make pipeline stable.

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 8]
```diff
@@ -0,0 +1,1192 @@
...
+    if (context->getSettingsRef().select_sequential_consistency)
+    {
+        if (const auto * replicated = dynamic_cast<const StorageReplicatedMergeTree *>(&reading->getMergeTreeData()))
+            return std::make_shared<PartitionIdToMaxBlock>(replicated->getMaxAddedBlocks());
+    }
+
+    return {};
+}
+
+static bool analyzeProjectionCandidate(
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:700]: Consider renaming to `analyzeProjectionCandidateParts`

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 9]
```diff
@@ -0,0 +1,1192 @@
...
+    /// Actions which need to be applied to columns from projection
+    /// in order to get all the columns required for aggregation.
+    ActionsDAGPtr dag;
+};
+
+/// Normal projection analysis result in case it can be applied.
+/// For now, it is empty.
+/// Normal projection can be used only if it contains all required source columns.
+/// It would not be hard to support pre-computed expressions and filtration.
+struct NormalProjectionCandidate : public ProjectionCandidate
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:256]: `OrdinaryProjectionCandidate`

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp (hunk 8 above)
Comment (novikd) [reply src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:700]: Also consider adding a comment

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 10]
```diff
@@ -0,0 +1,1192 @@
...
+    const Names & required_column_names,
+    const MergeTreeData::DataPartsVector & parts,
+    const StorageMetadataPtr & metadata,
+    const SelectQueryInfo & query_info,
+    const ContextPtr & context,
+    const std::shared_ptr<PartitionIdToMaxBlock> & max_added_blocks,
+    const ActionDAGNodes & added_filter_nodes)
+{
+    MergeTreeData::DataPartsVector projection_parts;
+    MergeTreeData::DataPartsVector normal_parts;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:713]: Rename to `ordinary_parts`

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 11]
```diff
@@ -0,0 +1,1192 @@
...
+{
+    IQueryPlanStep * step = node.step.get();
+    if (auto * reading = typeid_cast<ReadFromMergeTree *>(step))
+        return &node;
+
+    if (node.children.size() != 1)
+        return nullptr;
+
+    if (typeid_cast<ExpressionStep *>(step) || typeid_cast<FilterStep *>(step))
+        return findReadingStep(*node.children.front());
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:772]: Let's do it without recursion.

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 12]
```diff
@@ -0,0 +1,1192 @@
...
+        return true;
+    }
+};
+
+/// Required analysis info from aggregate projection.
+struct AggregateProjectionInfo
+{
+    ActionsDAGPtr before_aggregation;
+    NamesAndTypesList keys;
+    AggregateDescriptions aggregates;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:221]: Add a comment

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 13]
```diff
@@ -0,0 +1,1192 @@
...
+        return false;
+
+    if (!canUseProjectionForReadingStep(reading))
+        return false;
+
+    std::shared_ptr<PartitionIdToMaxBlock> max_added_blocks = getMaxAddedBlocks(reading);
+
+    auto candidates = getAggregateProjectionCandidates(node, *aggregating, *reading, max_added_blocks);
+
+    AggregateProjectionCandidate * best_candidate = nullptr;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:826]: Do we add only one projection per `optimizeUseAggProjections` call?

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 14]
```diff
@@ -0,0 +1,1192 @@
...
+    }
+
+    if (!best_candidate)
+        return false;
+
+    QueryPlanStepPtr projection_reading;
+    bool has_nornal_parts;
+
+    /// Add reading from projection step.
+    if (candidates.minmax_projection)
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:864]: Quite a tricky place to read. If it's not null, then `best_candidate` is set to `candidates.minmax_projection`.

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 15]
```diff
@@ -0,0 +1,1192 @@
...
+
+        if (best_candidate == nullptr || best_candidate->sum_marks > candidate.sum_marks)
+            best_candidate = &candidate;
+    }
+
+    if (!best_candidate)
+        return false;
+
+    QueryPlanStepPtr projection_reading;
+    bool has_nornal_parts;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:861]: `has_nornal_parts` -> `has_ordinary_parts`

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 16]
```diff
@@ -0,0 +1,1192 @@
...
+        iter = next;
+    }
+
+    if (iter == stack.rbegin())
+        return false;
+
+    const auto metadata = reading->getStorageMetadata();
+    const auto & projections = metadata->projections;
+
+    std::vector<const ProjectionDescription *> normal_projections;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:1029]: `ordinary_projections`

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 17]
```diff
@@ -0,0 +1,1192 @@
...
+    if (normal_projections.empty())
+        return false;
+
+    NormalQueryDAG query;
+    {
+        if (!query.build(*iter->node->children.front()))
+            return false;
+    }
+
+    std::list<NormalProjectionCandidate> candidates;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:1043]: Why `std::list`?

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 18]
```diff
@@ -0,0 +1,1192 @@
...
+
+        ActionDAGNodes added_filter_nodes;
+        if (query.filter_node)
+            added_filter_nodes.nodes.push_back(query.filter_node);
+
+        bool analyzed = analyzeProjectionCandidate(
+            candidate, *reading, reader, required_columns, parts,
+            metadata, query_info, context, max_added_blocks, added_filter_nodes);
+
+        if (!analyzed)
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:1076]: Looks strange ;) Better to rename.

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp [hunk 19]
```diff
@@ -0,0 +1,1192 @@
...
+        best_candidate->merge_tree_projection_select_result_ptr,
+        reading->isParallelReadingEnabled());
+
+    if (!projection_reading)
+    {
+        Pipe pipe(std::make_shared<NullSource>(proj_snapshot->getSampleBlockForColumns(required_columns)));
+        projection_reading = std::make_unique<ReadFromPreparedSource>(std::move(pipe));
+    }
+
+    bool has_nornal_parts = best_candidate->merge_tree_normal_select_result_ptr != nullptr;
```
Comment (novikd) [inline src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp:1124]: How can it be `false`?

**Review by novikd — COMMENTED** [review]
In general LGTM. I'd recommend splitting `optimizeUseProjections.cpp` into 2 files (to separate ordinary projections from aggregate projections) to improve readability.

File: tests/queries/0_stateless/02343_aggregation_pipeline.reference [hunk 20]
```diff
@@ -92,12 +92,15 @@ ExpressionTransform × 16
   (MergingAggregated)
   Resize 1 → 16
     MergingAggregatedTransform
-      Resize 2 → 1
+      Resize 17 → 1
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply tests/queries/0_stateless/02343_aggregation_pipeline.reference:95]: Cause 17 is 16 + 1, 16 is max_threads, and currently, we add resize after MergingAggregated step to improve performance.
https://github.com/ClickHouse/ClickHouse/commit/f5d26572df57a538486069557c214f43d51e442f#diff-2d416516943e820a31c0f539449b2f1586a798b6b4ec280772043337cc3334f6R69

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp [hunk 21]
```diff
@@ -1475,6 +1511,18 @@ bool ReadFromMergeTree::isQueryWithFinal() const
         return select.final();
 }
 
+bool ReadFromMergeTree::isQueryWithSampling() const
+{
+    if (context->getSettingsRef().parallel_replicas_count > 1 && data.supportsSampling())
+        return true;
+
+    const auto & select = query_info.query->as<ASTSelectQuery &>();
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/ReadFromMergeTree.cpp:1522]: Here I am checking table_expression_modifiers from analyzer first...

File: src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp (hunk 16 above)
Comment (KochetovNicolai) [reply src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp]: Currently, we call them normal. And have a special `ProjectionDescription::Type::Normal`.
I won't change this naming in this pr

**Pull request merged**